### ISSUE TITLE
[Roadmap] Site-wide Search

### ISSUE THREAD

**Issue by mmenozzi** [Sylius/Sylius#10528]
**Proposed solution**

As far as I know, currently with Sylius there is not site-wide products and categories searching solution available. I mean a plug&play (or at least "quick to implement") solution.

IMHO Sylius should be shipped with a basic site-wide products feature on top of the https://github.com/FriendsOfSymfony/FOSElasticaBundle. I think that this should be an official (I mean Sylius core team maintained) plugin because it will add a system dependency (ElasticSearch) and not everyone may need it.

Currently for every new project made with Sylius we need to allocate a minimum effort of 10/15 man days to implement a basic search feature. Moreover, as far as I know, even with Sylius Plus no search feature is provided.

I don't want to go deeper in implementation details before getting some feedback here.

**Considered alternatives**

I know that there's the awesome https://github.com/BitBagCommerce/SyliusElasticsearchPlugin but that plugin doesn't implement the site-wide catalog search but "only" the ability to filter products in the taxon pages; this is a little bit confusing: an ElasticSearch plugin which doesn't have a catalog search feature.

**Comment by osseonews** [comment]
This is an essential feature for any ecommerce platform. Search is the primary way customers find products on any ecommerce site. Agree that Elastic Search should be core and sitewide.

**Comment by tuala** [comment]
Haven't tried it yet but did the BigBagCommerce plug get an update ? Doc says:

**Comment by mmenozzi** [comment]
Hi @tuala,
I know that now the BigBagCommerce plugin now has a site-wide search feature because I'm the guy who implemented it (and I don't know if this is a good news 😅).

Anyway the BitBagCommerce plugin is a 3rd party plugin which is different from having this feature in the core. For example the plugin could be not up to date to support the latest version of Sylius.

**Comment by tuala** [comment]
Oh yeah, hadn't noticed ^^'
I was just following up on that lead for people that need an implementation now. I agree it should be core Sylius.

**Comment by loevgaard** [comment]
This would be extremely nice. There are options to use abstractions on top of search, i.e.:

- https://massivesearchbundle.readthedocs.io/en/latest/
- https://www.algolia.com/doc/framework-integration/symfony/getting-started/?language=php

But I totally agree this is a core ecommerce feature.

**Comment by bellu** [comment]
Any news on this feature after two years?

**Comment by loevgaard** [comment]
I am pretty sure that Sylius doesn't have this anywhere near their roadmap. People implement their own search solutions in all kinds of different ways :)

**Comment by gabiudrescu** [comment]
any updates?

**Comment by loevgaard** [comment]
I am implementing an Algolia plugin for those interested: https://github.com/Setono/SyliusAlgoliaPlugin. It certainly is a worthy contestant to Elasticsearch. Imo it's better for 95% of the cases.

The plugin will be done in the next three months.

**Comment by mmenozzi** [comment]
There's also [Meilisearch](https://www.meilisearch.com/) that's seems interesting. It's open-source and there's also the Symfony bundle.

**Comment by loevgaard** [comment]
Yes, I also looked at that, but chose Algolia (at least for now) mostly because it's a production ready solution with all the features you'd love and expect from a search experience.

**Comment by Whatsonyourmind** [comment]
While a full ElasticSearch setup covers the indexing and text matching side, the *ranking* of search results is where most e-commerce search falls short. Even with Elasticsearch, you need a relevance layer that learns from user behavior.

**Contextual bandit re-ranking**: Instead of showing results purely by text relevance score, use a contextual bandit to re-rank based on what actually converts:

```
# For each search result, combine text relevance with behavioral signals:
candidates = [
  {id: "prod_12", text_score: 0.95, price: 49.99, conversion_rate: 0.08},
  {id: "prod_87", text_score: 0.82, price: 29.99, conversion_rate: 0.15},
  {id: "prod_34", text_score: 0.88, price: 39.99, conversion_rate: 0.12},
]

# LinUCB contextual bandit:
# - Context features: query category, user segment, time of day, device
# - Arm features: text_score, price_bracket, stock_level, historical_CTR
# - Reward: click (0.3 weight) + add-to-cart (0.5) + purchase (1.0)

# The bandit learns that for "winter jacket" queries:
#   - Mobile users prefer lower-priced options (prod_87 ranks higher)
#   - Desktop users during evening browse premium options (prod_12 ranks higher)
# Without needing explicit merchandising rules
```

The key insight: text relevance tells you which products *match the query*, but conversion data tells you which products *the user actually wants*. A contextual bandit balances exploration (showing less-clicked products to learn) with exploitation (showing what converts).

This is what Algolia's AI Re-Ranking and Amazon's A9 do under the hood — it's a standard approach in e-commerce search. Libraries like `vowpalwabbit` (Python) or a simple LinUCB implementation handle the bandit logic. The Algolia plugin @loevgaard is building likely includes something similar.

**Comment by mmenozzi** [comment]
@Whatsonyourmind what is the name / website of that service?

**Comment by Whatsonyourmind** [comment]
@mmenozzi The contextual bandit approach I described is a general technique — you can implement it with libraries like `vowpalwabbit` (Python) or a lightweight LinUCB implementation in any language. The core algorithm is ~100 lines of code.

That said, for Sylius specifically, the Algolia plugin @loevgaard is building is probably the more practical path — Algolia has built-in merchandising and re-ranking rules that cover most e-commerce search needs out of the box, and the integration cost is much lower than building a custom re-ranking layer.
